### PR TITLE
[MCOL-563] Disk-based aggregation

### PR DESCRIPTION
**Pull request by mariadb-AlexeyAntipovsky**

### CODE REVIEW

**Comment by davidhall586** [comment]
Please never merge this. This feature should not be added to develop-5

File: tools/rgprint/rgprint.cpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
+//
```
Comment (drrtuy) [inline tools/rgprint/rgprint.cpp]: License.

File: writeengine/splitter/we_splitterapp.cpp [hunk 2]
```diff
@@ -486,6 +486,10 @@ void WESplitterApp::invokeCpimport()
     {
         //we need something that works on Windows as well as linux
         v2[i++] = arg;
+        if (i >= v2.size())
```
Comment (drrtuy) [inline writeengine/splitter/we_splitterapp.cpp]: Could you pull this in a separate patch?

File: tools/rgprint/rgprint.cpp (hunk 1 above)
Comment (mariadb-AlexeyAntipovsky) [reply tools/rgprint/rgprint.cpp]: Forgot about it. Actually it's just debugging tool, so I've also removed it from tools/CMakeLists.txt

File: writeengine/splitter/we_splitterapp.cpp (hunk 2 above)
Comment (mariadb-AlexeyAntipovsky) [reply writeengine/splitter/we_splitterapp.cpp]: https://github.com/mariadb-corporation/mariadb-columnstore-engine/pull/1847
https://github.com/mariadb-corporation/mariadb-columnstore-engine/pull/1848

File: dbcon/joblist/tupleaggregatestep.h [hunk 3]
```diff
@@ -156,14 +157,37 @@ class TupleAggregateStep : public JobStep, public TupleDeliveryStep
         {}
         void operator()()
         {
-            utils::setThreadName("TASThrAggr");
+            std::string t{"TASThrAggr"};
+            t.append(std::to_string(fThreadID));
+            utils::setThreadName(t.c_str());
+            //utils::setThreadName("TASThrAggr");
```
Comment (drrtuy) [inline dbcon/joblist/tupleaggregatestep.h]: Plz remove the commented line.

File: dbcon/joblist/tupleaggregatestep.cpp [hunk 4]
```diff
@@ -366,16 +366,18 @@ TupleAggregateStep::TupleAggregateStep(
...
+    fNumOfBuckets = fRm->aggNumBuckets();
+    fNumOfRowGroups = fRm->aggNumRowGroups();
+
     fAggregator->setInputOutput(fRowGroupIn, &fRowGroupOut);
 
     // decide if this needs to be multi-threaded
     RowAggregationDistinct* multiAgg = dynamic_cast<RowAggregationDistinct*>(fAggregator.get());
     fIsMultiThread = (multiAgg || fAggregator->aggMapKeyLength() > 0);
 
     // initialize multi-thread variables
```
Comment (drrtuy) [inline dbcon/joblist/tupleaggregatestep.cpp]: Take a comment with the asignments you moved.

File: dbcon/joblist/tupleaggregatestep.cpp [hunk 5]
```diff
@@ -5192,6 +5194,25 @@ void TupleAggregateStep::aggregateRowGroups()
     }
 }
 
+void TupleAggregateStep::threadedAggregateFinalize(uint32_t threadID)
+{
+  for (uint32_t i = 0; i < fNumOfBuckets; ++i)
+  {
+    if (fAgg_mutex[i]->try_lock())
```
Comment (drrtuy) [inline dbcon/joblist/tupleaggregatestep.cpp:5204]: I am curios about the reasons why try_lock() and not lock() is used here?

File: utils/rowgroup/rowstorage.h [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+  void freeData();
+
+  /** @brief Move internal data & row position inside [insIdx, startIdx] up by 1.
+   *
+   * @param startIdx(in) last element's index to move
+   * @param insIdx(in)   first element's index to move
+   */
+  void shiftUp(size_t startIdx, size_t insIdx);
+
+  inline static size_t mixHash(size_t x) noexcept
```
Comment (drrtuy) [inline utils/rowgroup/rowstorage.h]: I am curios which hash function mixer did you re-use or it was your own and why did you choose uint64_t and not uint32_t as the hash output?

File: utils/rowgroup/rowstorage.cpp [hunk 7]
```diff
@@ -0,0 +1,2139 @@
...
+      h.add(r.getCharset(i), r.getConstString(i));
+      break;
+
+    default:
+      ret = hashData(r.getData() + r.getOffset(i), r.getColumnWidth(i), ret);
+      break;
+    }
+  }
+
+  if (UNLIKELY(h.wasUsed()))
```
Comment (drrtuy) [inline utils/rowgroup/rowstorage.cpp]: I don't think this condition is unlikely for a whole Row.

File: dbcon/joblist/tupleaggregatestep.h (hunk 3 above)
Comment (mariadb-AlexeyAntipovsky) [reply dbcon/joblist/tupleaggregatestep.h]: Oh, sorry, I missed this. Actually I added thr number to the thread name for debugging (so I was able to see exactly which thread was working on particular bucket). Do you thing it makes sense for production or should I rollback this?

File: dbcon/joblist/tupleaggregatestep.cpp (hunk 4 above)
Comment (mariadb-AlexeyAntipovsky) [reply dbcon/joblist/tupleaggregatestep.cpp]: This is accidentally left over from the experiments, so I'll rollback it.

File: dbcon/joblist/tupleaggregatestep.cpp (hunk 5 above)
Comment (mariadb-AlexeyAntipovsky) [reply dbcon/joblist/tupleaggregatestep.cpp:5204]: We go through the bucket list and take the first one that isn't currently being finalized, so there is no need to wait for the processing to complete. The state of the race here is resolved by the flag in RowAggStorage.

File: utils/rowgroup/rowstorage.cpp (hunk 7 above)
Comment (mariadb-AlexeyAntipovsky) [reply utils/rowgroup/rowstorage.cpp]: Yes, you're right.

File: utils/rowgroup/rowaggregation.cpp [hunk 8]
```diff
@@ -751,9 +680,47 @@ void RowAggregation::initialize()
...
+      {
+        allow_gen = false;
+        break;
+      }
+    }
+
+    config::Config* config = config::Config::makeConfig();
+    string tmpDir = config->getConfig("RowAggregation", "TempDir");
+    if (tmpDir.empty() || tmpDir == "/")
+      tmpDir.assign("/tmp/cs-agg");
```
Comment (drrtuy) [inline utils/rowgroup/rowaggregation.cpp]: A questionable default since we want to offload the data from RAM and /tmp in most cases is tmpfs. I think it would be better to check if disk-based aggregation is enable and check if RowAggregation.TempDir is set. If not we should throw an error.

File: utils/rowgroup/rowaggregation.cpp [hunk 9]
```diff
@@ -4180,7 +4079,14 @@ void RowAggregationUMP2::doAvg(const Row& rowIn, int64_t colIn, int64_t colOut,
     if (cnt == 0)
     {
         fRow.setLongDoubleField(valIn, colOut);
-        fRow.setUintField(rowIn.getUintField(colIn + 1), colAux);
+        if (merge)
```
Comment (drrtuy) [inline utils/rowgroup/rowaggregation.cpp]: What is the difference b/w these two branches?

File: utils/rowgroup/rowstorage.h [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+  std::string makeDumpFilename(int32_t gen = -1) const;
+
+private:
+  static constexpr size_t   INIT_SIZE{sizeof(uint64_t)};
+  static constexpr uint32_t INIT_INFO_BITS{5};
+  static constexpr uint8_t  INIT_INFO_INC{1U << INIT_INFO_BITS};
+  static constexpr size_t   INFO_MASK{INIT_INFO_INC - 1U};
+  static constexpr uint8_t  INIT_INFO_HASH_SHIFT{0};
+
+  RowPosHashStorage* fHashes{nullptr};
```
Comment (drrtuy) [inline utils/rowgroup/rowstorage.h:291]: Just a note that this type of initialization breaks exuberant-ctags shipped with U20.

File: utils/loggingcpp/ErrorMessage.txt [hunk 11]
```diff
@@ -100,6 +100,10 @@
 
 2053	ERR_FUNC_OUT_OF_RANGE_RESULT	The result is out of range for function %1% using value(s): %2% %3%
 
+2054	ERR_DISKAGG_ERROR	Unknown error while aggregation.
```
Comment (drrtuy) [inline utils/loggingcpp/ErrorMessage.txt:103]: There is no way to detect which method throws this error. The thow should log at least a method name.

File: oam/etc/Columnstore.xml.singleserver [hunk 12]
```diff
@@ -516,6 +516,7 @@
                <!-- <RowAggrThreads>4</RowAggrThreads> --> <!-- Default value is the number of cores -->
 		<!-- <RowAggrBuckets>32</RowAggrBuckets> --> <!-- Default value is number of cores * 4 -->
 		<!-- <RowAggrRowGroupsPerThread>20</RowAggrRowGroupsPerThread> --> <!-- Default value is 20 -->
+		<!-- <AllowDiskBasedAggregation>N</AllowDiskBasedAggregation> --> <!-- Default value is N -->
```
Comment (drrtuy) [inline oam/etc/Columnstore.xml.singleserver:519]: JFYI. This file is obsolete and should be removed but not with this patch.

**Review by drrtuy — COMMENTED** [review]
I've no structural suggestions so far. My main question is why RowHashPos uses uint64_t for its hash and not uint32_t?

File: utils/rowgroup/rowstorage.h (hunk 6 above)
Comment (mariadb-AlexeyAntipovsky) [reply utils/rowgroup/rowstorage.h]: mixHash was left over after using the original Row::hash hasher, I removed it. robinhood hash looks more evenly distributed.
And 64 bits for hash was chosen simply because RH uses 64 bits.

File: utils/rowgroup/rowaggregation.cpp (hunk 9 above)
Comment (mariadb-AlexeyAntipovsky) [reply utils/rowgroup/rowaggregation.cpp]: If it was used while update the aggregated row with the input row it just increments the aggregated value by 1, otherwise it sums the target aggregated value with the source aggregated value.

File: utils/rowgroup/rowstorage.h (hunk 10 above)
Comment (mariadb-AlexeyAntipovsky) [reply utils/rowgroup/rowstorage.h:291]: It seems that universal-ctags is able to work with modern syntax

Comment (drrtuy) [reply utils/rowgroup/rowstorage.h:291]: Thx for the note.

File: utils/rowgroup/rowaggregation.cpp (hunk 9 above)
Comment (drrtuy) [reply utils/rowgroup/rowaggregation.cpp]: In this particular case the statements in both scopes are identical. Or I am wrong?

**Review by drrtuy — APPROVED** [review]
